### PR TITLE
CLI rewrite

### PR DESCRIPTION
**Pull request by notfilippo**
### Rewrite of existing code

A rewrite is considered finished if works as expected (duh), has comments, passes `lint` and `fmt` and (for the most part) has tests.

- [x] logger
- [x] configuration
- [x] keyfile
- [x] version
- [x] file handling
- [x] commands
    - [x] init
    - [x] link
    - [x] publish
    - [x] upgrade

### CODE REVIEW

**Comment by zorbyte** [comment]
Wait on the JSDoc, we have an idea for this later.

File: src/module.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  constructor(module: IModule) {
+    this.name = module.name;
+    this.owner = module.owner;
+    this.description = module.description;
+    this.latestVersion = module.latestVersion;
+    this.latestStableVersion = module.latestStableVersion;
+    this.packageUploadNames = module.packageUploadNames;
+  }
+
+  getLatestVersion(stable: boolean): string {
```
Comment (SteelAlloy) [inline src/module.ts]:
```suggestion
  getLatestVersion(stable?: boolean): string {
```
Without this I was getting the following:
```ts
error: TS2345 [ERROR]: Argument of type 'boolean | undefined' is not assignable to parameter of type 'boolean'.
  Type 'undefined' is not assignable to type 'boolean'.
    latest = existing.getLatestVersion(egg.stable);
                                       ~~~~~~~~~~
    at file:///P:/GitHub/eggs/src/commands/publish.ts:181:40
```

File: src/api/fetch.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import { ENDPOINT } from "./common.ts";
+
+export async function fetchResource<T>(query: string): Promise<T | undefined> {
+  // TODO(@qu4k): add test resource
+  try {
+    const response = await fetch(`${ENDPOINT}${query}`);
+    if (!response || !response.ok) return undefined;
+    const value = await response.json();
+    return value as T;
+  } catch (_) {
```
Comment (SteelAlloy) [inline src/api/fetch.ts]:
```suggestion
  } catch {
```
This is not needed

File: src/api/post.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        "Content-Type": "application/json",
+        ...headers,
+      },
+      body: JSON.stringify(data),
+    });
+
+    if (!response || !response.ok) return undefined;
+    const value = await response.json();
+    return value as T;
+  } catch (_) {
```
Comment (SteelAlloy) [inline src/api/post.ts]:
```suggestion
  } catch {
```

File: src/commands/init.ts [hunk 4]
```diff
@@ -1,56 +1,68 @@
...
+  readConfig,
+  writeConfig,
+} from "../config.ts";
 import { version } from "../version.ts";
 
-function detectConfig(): ConfigFormats {
-  if (pathExists("egg.yaml")) return "yaml";
-  else if (pathExists("egg.yml")) return "yml";
-  return "json";
+/**
```
Comment (SteelAlloy) [inline src/commands/init.ts]: As we try to follow Deno style guide, this includes
[Vertical spacing should be minimized whenever possible.](https://github.com/denoland/deno/blob/4ebd24342368adbb99582b87dc6c4b8cb6f44c87/docs/contributing/style_guide.md#use-jsdoc-for-exported-symbols)
```ts
/** Init Command.
 * `init` creates (or overrides) configuration in
 * the cwd with an interactive prompt. */
```
The same goes for the other comments

File: src/commands/init.ts [hunk 5]
```diff
@@ -1,56 +1,68 @@
...
+    maxLength: 4294967295,
+  });
+  const stable: boolean = await Confirm.prompt({
+    message: "Is this a stable version?",
+    default: currentConfig.stable,
+  });
+  const files: string[] = await List.prompt(
+    "Enter the files and relative directories that nest.land will publish separated by a comma.",
+  );
+  const format: string = await Select.prompt({
```
Comment (SteelAlloy) [inline src/commands/init.ts:44]: Thank you for making this change!

**Review by SteelAlloy — CHANGES_REQUESTED** [review]
Quality work
Just a few minor changes and I'm good with it.

File: src/api/fetch.ts (hunk 2 above)
Comment (notfilippo) [reply src/api/fetch.ts]: TIL catch parameter is optional, thx

File: src/api/post.ts (hunk 3 above)
Comment (notfilippo) [reply src/api/post.ts]: ditto

File: src/commands/init.ts (hunk 4 above)
Comment (notfilippo) [reply src/commands/init.ts]: Thanks for pointing this out

File: src/commands/init.ts (hunk 5 above)
Comment (notfilippo) [reply src/commands/init.ts:44]: :)

**Pull request merged**